src/txHashBuilder.c

```diff
@@ -288,20 +292,358 @@ void txHashBuilder_addCertificate_delegation(
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS);
+	ASSERT(builder->poolCertificateData.remainingOwners > 0);
+	builder->poolCertificateData.remainingOwners--;
+
+	// Bytes[poolKeyHash]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, stakingKeyHashSize);
+		BUILDER_APPEND_DATA(stakingKeyHash, stakingKeyHashSize);
+	}
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_enterRelays(tx_hash_builder_t* builder)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_enterRelays: %d", builder->state);
+
+	// enter empty owners if none were received
+	if (builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL)
+		txHashBuilder_addPoolRegistrationCertificate_enterOwners(builder);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS);
+	ASSERT(builder->poolCertificateData.remainingOwners == 0);
+
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, builder->poolCertificateData.remainingRelays);
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS;
+}
+
+static void addRelay_updateState(tx_hash_builder_t* builder)
+{
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL:
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS:
+		txHashBuilder_addPoolRegistrationCertificate_enterRelays(builder);
+		break;
+
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS:
+		break; // we want to be here
+
+	default:
+		ASSERT(false);
+	}
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addRelay0(
+        tx_hash_builder_t* builder,
+        const uint16_t* port,
+        const uint8_t* ipv4, size_t ipv4Size,
+        const uint8_t* ipv6, size_t ipv6Size
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addRelay0: %d", builder->state);
+
+	addRelay_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+	ASSERT(builder->poolCertificateData.remainingRelays > 0);
+	builder->poolCertificateData.remainingRelays--;
+
+	ASSERT(ipv4Size < BUFFER_SIZE_PARANOIA);
+	ASSERT(ipv6Size < BUFFER_SIZE_PARANOIA);
+
+	// Array(4)[
+	//   Unsigned[0]
+	//   Unsigned[port] / Null
+	//   Bytes[ipv4] / Null
+	//   Bytes[ipv6] / Null
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 4);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 0);
+		}
+		{
+			if (port != NULL) {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, *port);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			if (ipv4 != NULL) {
+				ASSERT(ipv4Size == 4);
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, ipv4Size);
+				BUILDER_APPEND_DATA(ipv4, ipv4Size);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			if (ipv6 != NULL) {
+				ASSERT(ipv6Size == 16);
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, ipv6Size);
+
+				// serialized as 4 big-endian uint32
+				for (size_t i = 0; i < 4; i++) {
+					for (size_t j = 0; j < 4; j++) {
```

Comment:
You should preferably use endian-conversion utilities we have, not this hand-written code

Reply:
Something like
```
					uint8_t chunk[4];
					uint32_t* valuePtr = (uint32_t *)(ipv6 + 4 * i);
					u4be_write(chunk, *valuePtr);
```
?